File: compiler/queries/placeholders.go
package queries

import (
	"wipple/database"
	"wipple/nodes/expressions"
	"wipple/typecheck"
)

func Placeholder(db *database.Db, node database.Node, filter func(node database.Node) bool, f func(typecheck.Type)) {
	if _, ok := database.GetFact[expressions.IsPlaceholder](node); ok {
		if fact, ok := database.GetFact[typecheck.TypedFact](node); ok && fact.Group != nil && len(fact.Group.Types) > 0 {
			f(fact.Group.Types[0])
		} else {
			f(nil)
		}
	}
}

File: compiler/src/queries/placeholders.rs
use crate::{
    database::NodeRef,
    nodes::IsPlaceholder,
    queries::QueryCtx,
    typecheck::{ConstructedType, Typed},
};

pub fn placeholder(ctx: &QueryCtx<'_>, f: &mut dyn FnMut((NodeRef, Option<ConstructedType>))) {
    if !ctx.db.contains::<IsPlaceholder>(&ctx.node) {
        return;
    }

    let ty = ctx
        .db
        .get::<Typed>(&ctx.node)
        .and_then(|Typed { group }| group)
        .and_then(|group| group.types.first().cloned());

    f((ctx.node.clone(), ty));
}
